File: Python/ByRound/1716/1716_D_Chip_Move.py
# Problem: CF 1716 D - Chip Move
# https://codeforces.com/contest/1716/problem/D

"""
Purpose: 
    Count the number of ways to reach each point x in [1, n] starting from 0,
    where each move's length must be divisible by k, k+1, k+2, etc. for consecutive moves.
    
Algorithm:
    - Use dynamic programming with modular arithmetic.
    - For each step, maintain a sliding window of possible previous positions.
    - Use a memoization array to track how many ways we can reach each modulo class.
    
Time Complexity: O(n * k)
Space Complexity: O(n)

The approach uses a rolling window technique:
- At each iteration, we compute the number of ways to reach positions up to L + k - 1.
- The memo array tracks how many paths end at each modulo class.
- We update the answer array and memo array iteratively.
"""

mod = 998244353
n, k = map(int, input().split())

L = 0
ans = [0] * (n + 1)
now = [0] * (n + 1)
now[0] = 1
while True:
    memo = [0] * k
    for i in range(L, n + 1):
        # Add the number of ways to reach position i from previous positions
        ans[i] += memo[i % k]
        ans[i] %= mod
        # Update memo and now arrays for next iteration
        memo[i % k], now[i] = (memo[i % k] + now[i]) % mod, memo[i % k]
    L += k
    k += 1
    if L > n:
        break

print(*ans[1:])


# https://github.com/VaHiX/CodeForces/

File: CPP/ByTier/D/1716_D_Chip_Move.cpp
// Problem: CF 1716 D - Chip Move
// https://codeforces.com/contest/1716/problem/D

/*
Purpose: This code counts the number of ways to reach each point from 1 to n on a coordinate line,
         where each move's length must be divisible by k, k+1, k+2, etc. for the 1st, 2nd, 3rd... moves.
         The solution uses dynamic programming with a sliding window technique.

Algorithms/Techniques: 
- Dynamic Programming with rolling array optimization
- Modular arithmetic to prevent overflow
- Precomputation of dp states up to a fixed limit (650)

Time Complexity: O(n * sqrt(n)) - The loop runs up to sqrt(n) times, and for each iteration, 
                                 we process up to n elements. In practice, it's more efficient 
                                 due to the early termination condition.

Space Complexity: O(n) - Only two rows of dp array and ans array are used, both of size n.

*/
#include <stdio.h>
#include <iostream>

int n, k, dp[2][200005] = {1}, ans[200005]; // dp is the rolling array, ans stores final results
int main() {
  std::cin >> n >> k;
  for (int i = 1; i <= 650; i++) // Loop up to a fixed threshold to compute dp values
    for (int j = 0; j <= n; j++)
      ans[j] = (ans[j] + (dp[i & 1][j] = j < i + k - 1 // Compute current dp state
                                             ? 0
                                             : (dp[i & 1][j - i - k + 1] + // Add previous dp states
                                                dp[i & 1 ^ 1][j - i - k + 1]) %
                                                   998244353)) %
               998244353; // Apply modulo to avoid overflow
  for (int i = 1; i <= n; i++)
    printf("%d ", ans[i]); // Output the results
}


// https://github.com/VaHiX/CodeForces/